drv/lpc55-update-server/src/main.rs

```diff
@@ -226,6 +234,203 @@ impl idl::InOrderUpdateImpl for ServerImpl {
+                //
+                // That means you can apply this operation several times before
+                // resetting without burning many monotonic versions, if you
+                // want to do that for some reason.
+                //
+                // The addresses of these pages are as follows (see Figure 13,
+                // "Protected Flash Region," in UM11126 rev 2.4, or the NXP
+                // flash layout spreadsheet):
+                //
+                // Page     Addr        16-byte word number
+                // Scratch  0x9_DE00    0x9DE0
+                // Ping     0x9_E000    0x9E00
+                // Pong     0x9_E200    0x9E20
+
+                let cfpa_word_number = {
+                    // Read the two versions. We do this with smaller buffers so
+                    // we don't need 2x 512B buffers to read the entire CFPAs.
+                    let mut ping_header = [0u32; 4];
+                    let mut pong_header = [0u32; 4];
+
+                    indirect_flash_read(
+                        &mut self.flash,
+                        0x9E00,
+                        core::slice::from_mut(&mut ping_header),
+                    )?;
+                    indirect_flash_read(
+                        &mut self.flash,
+                        0x9E20,
+                        core::slice::from_mut(&mut pong_header),
+                    )?;
+
+                    // Work out where to read the authoritative contents from.
+                    if ping_header[1] >= pong_header[1] {
+                        0x9E00
+                    } else {
+                        0x9E20
+                    }
+                };
+
+                // Read current CFPA contents.
+                let mut cfpa = [[0u32; 4]; 512 / 16];
+                indirect_flash_read(
+                    &mut self.flash,
+                    cfpa_word_number,
+                    &mut cfpa,
+                )?;
+
+                // Increment the monotonic version. The manual doesn't specify
+                // how the version numbers are compared or what happens if they
+                // wrap, so, we'll treat wrapping as an error and report it for
+                // now. (Note that getting this version to wrap _should_ require
+                // more write cycles than the flash can take.)
+                let new_version =
+                    cfpa[0][1].checked_add(1).ok_or(UpdateError::SecureErr)?;
+                cfpa[0][1] = new_version;
+                // Alter the boot setting. The boot setting (per RFD 374) is in
+                // the lowest bit of the 32-bit word starting at (byte) offset
+                // 0x100. This is flash word offset 0x10.
+                //
+                // Leave remaining bits undisturbed; they are currently
+                // reserved.
+                cfpa[0x10][0] &= !1;
+                cfpa[0x10][0] |= if slot == SlotId::A { 0 } else { 1 };
```

Comment:
Just thinking this through. Looks good.

Physical access is out of scope for the threat model including glitching attacks, so using a single bit for bank selection is likely not a problem.

Unlike the transient boot preference selection (Once) in RAM, the CFPA-based persistent selection (Forever) is ok to be a single bit because there is ECC protection for CFPA storage (true?) and/or the trailing SHA2-256 hash.

There are three states, a bank has never been selected, A selected, B selected.
The never-been-selected state is only for a short while, if ever, while being manufactured.
Additionally, this is just a preference, if the preferred bank is not suitable for boot, no attempt
will be made to run it.

Reply:
*nod* That's basically my reasoning. I discussed this in an aside in [the data storage section of RFD374](https://rfd.shared.oxide.computer/rfd/374#_cfpa_data_storage_specifics) (sorry, non-Oxide viewers). The CFPA storage does indeed have _both_ ECC _and_ the trailing SHA256; the ECC is the actual protection, the SHA256 is mostly just a handshake with the ROM.

The main alternative I considered was to have a fancier bool that had possibly invalid states, so you could detect garbage (the "secure bool" pattern). But if we have a fancy bool that admits invalid states, we need to define a default behavior for what happens if the CFPA is found in an illegal state. There are three options as I see it:
1. Ignore it and boot slot A
2. Ignore it and boot slot B
3. Brick the device.

I don't like (3), so I rejected it. At that point, if the fallback position is going to be booting slot A or B ... it's a lot like having a single bit, which is unambiguous.

So in the end,
- There's a single bit that controls the preference
- It is ECC protected but can be overwritten at runtime (in a controlled fashion)
- Manufacturing default state is zeroed and reads as "prefer A," so there isn't a "not yet selected" state
- Once the device is locked, the ROM protects this sector from unexpected erasure, so "erased" isn't a state we need to consider.